torchmetrics/functional/text/squadv2.py

```diff
@@ -0,0 +1,316 @@
+
+
+def compute_exact(a_gold, a_pred) -> int:
+    return int(normalize_answer(a_gold) == normalize_answer(a_pred))
+
+
+def compute_f1(a_gold, a_pred) -> float:
+    gold_toks = get_tokens(a_gold)
+    pred_toks = get_tokens(a_pred)
+    common = collections.Counter(gold_toks) & collections.Counter(pred_toks)
+    num_same = sum(common.values())
+    if len(gold_toks) == 0 or len(pred_toks) == 0:
+        # If either is no-answer, then F1 is 1 if they agree, 0 otherwise
+        return int(gold_toks == pred_toks)
+    if num_same == 0:
+        return 0
+    precision = 1.0 * num_same / len(pred_toks)
+    recall = 1.0 * num_same / len(gold_toks)
+    f1 = (2 * precision * recall) / (precision + recall)
+    return f1
+
+
+def get_raw_scores(dataset, preds) -> Tuple[Dict[Any, int], Dict[Any, float]]:
+    exact_scores = {}
+    f1_scores = {}
+    for article in dataset:
+        for p in article["paragraphs"]:
+            for qa in p["qas"]:
+                qid = qa["id"]
+                gold_answers = [t for t in qa["answers"]["text"] if normalize_answer(t)]
+                if not gold_answers:
+                    # For unanswerable questions, only correct answer is empty string
+                    gold_answers = [""]
+                if qid not in preds:
+                    print("Missing prediction for %s" % qid)
+                    continue
+                a_pred = preds[qid]
+                # Take max over all gold answers
+                exact_scores[qid] = max(compute_exact(a, a_pred) for a in gold_answers)
+                f1_scores[qid] = max(compute_f1(a, a_pred) for a in gold_answers)
+    return exact_scores, f1_scores
+
+
+def apply_no_ans_threshold(scores, na_probs, qid_to_has_ans, na_prob_thresh) -> Dict:
+    new_scores = {}
+    for qid, s in scores.items():
+        pred_na = na_probs[qid] > na_prob_thresh
+        if pred_na:
+            new_scores[qid] = float(not qid_to_has_ans[qid])
+        else:
+            new_scores[qid] = s
+    return new_scores
+
+
+def make_eval_dict(exact_scores, f1_scores, qid_list=None) -> OrderedDict[str, Union[Union[float, int], Any]]:
+    if not qid_list:
+        total = len(exact_scores)
+        return collections.OrderedDict([
+            ("exact", 100.0 * sum(exact_scores.values()) / total),
+            ("f1", 100.0 * sum(f1_scores.values()) / total),
+            ("total", total),
+        ])
+    else:
+        total = len(qid_list)
+        return collections.OrderedDict([
+            ("exact", 100.0 * sum(exact_scores[k] for k in qid_list) / total),
+            ("f1", 100.0 * sum(f1_scores[k] for k in qid_list) / total),
+            ("total", total),
+        ])
+
+
+def merge_eval(main_eval, new_eval, prefix) -> None:
+    for k in new_eval:
+        main_eval["%s_%s" % (prefix, k)] = new_eval[k]
+
+
+def plot_pr_curve(precisions, recalls, out_image, title) -> None:
+    plt.step(recalls, precisions, color="b", alpha=0.2, where="post")
```

Comment:
use fig as an anchor for plotting, this is very sensitive to actually active figure